Crash when opening Naming Numbers revision card in Place Values

Stack trace:

```
    java.lang.RuntimeException: Failed to call observer method
        at androidx.lifecycle.ClassesInfoCache$MethodReference.invokeCallback(ClassesInfoCache.java:226)
        at androidx.lifecycle.ClassesInfoCache$CallbackInfo.invokeMethodsForEvent(ClassesInfoCache.java:194)
        at androidx.lifecycle.ClassesInfoCache$CallbackInfo.invokeCallbacks(ClassesInfoCache.java:185)
        at androidx.lifecycle.ReflectiveGenericLifecycleObserver.onStateChanged(ReflectiveGenericLifecycleObserver.java:37)
        at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.java:361)
        at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.java:300)
        at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.java:339)
        at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.java:145)
        at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.java:131)
        at androidx.fragment.app.Fragment.performStart(Fragment.java:2732)
        at androidx.fragment.app.FragmentStateManager.start(FragmentStateManager.java:331)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1239)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1398)
        at androidx.fragment.app.FragmentManager.moveFragmentToExpectedState(FragmentManager.java:1476)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1541)
        at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:2858)
        at androidx.fragment.app.FragmentManager.dispatchStart(FragmentManager.java:2818)
        at androidx.fragment.app.FragmentController.dispatchStart(FragmentController.java:258)
        at androidx.fragment.app.FragmentActivity.onStart(FragmentActivity.java:550)
        at androidx.appcompat.app.AppCompatActivity.onStart(AppCompatActivity.java:201)
        at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1432)
        at android.app.Activity.performStart(Activity.java:7825)
        at android.app.ActivityThread.handleStartActivity(ActivityThread.java:3294)
        at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:221)
        at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:201)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:173)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.IndexOutOfBoundsException: setSpan (1966 ... 1965) has end before start
        at android.text.SpannableStringBuilder.checkRange(SpannableStringBuilder.java:1318)
        at android.text.SpannableStringBuilder.setSpan(SpannableStringBuilder.java:684)
        at android.text.SpannableStringBuilder.setSpan(SpannableStringBuilder.java:676)
        at android.text.HtmlToSpannedConverter.convert(Html.java:774)
        at android.text.Html.fromHtml(Html.java:243)
        at androidx.core.text.HtmlCompat.fromHtml(HtmlCompat.java:162)
        at org.oppia.util.parser.CustomHtmlContentHandler$Companion.fromHtml(CustomHtmlContentHandler.kt:146)
        at org.oppia.util.parser.HtmlParser.parseOppiaHtml(HtmlParser.kt:77)
        at org.oppia.util.parser.HtmlParser.parseOppiaHtml$default(HtmlParser.kt:48)
        at org.oppia.app.topic.revisioncard.RevisionCardViewModel.processExplanationResult(RevisionCardViewModel.kt:80)
        at org.oppia.app.topic.revisioncard.RevisionCardViewModel.access$processExplanationResult(RevisionCardViewModel.kt:22)
```

Seems related to the recent bullet span work.

--- a/utility/src/main/java/org/oppia/util/parser/BulletTagHandler.kt
+++ b/utility/src/main/java/org/oppia/util/parser/BulletTagHandler.kt
@@ -2,9 +2,8 @@ package org.oppia.util.parser
 
 import android.text.Editable
 import android.text.Spannable
-import android.text.SpannableStringBuilder
+import android.text.Spanned
 import android.text.style.BulletSpan
-import org.xml.sax.Attributes
 
 /** The custom tag corresponding to [BulletTagHandler]. */
 const val CUSTOM_BULLET_LIST_TAG = "oppia-li"
@@ -14,24 +13,21 @@ const val CUSTOM_BULLET_LIST_TAG = "oppia-li"
  * [CustomHtmlContentHandler].
  */
 class BulletTagHandler : CustomHtmlContentHandler.CustomTagHandler {
-  override fun handleTag(
-    attributes: Attributes,
-    openIndex: Int,
-    closeIndex: Int,
-    output: Editable
-  ) {
-    val spannableBuilder = SpannableStringBuilder(
-      output.subSequence(
-        openIndex,
-        closeIndex
-      )
-    )
-    spannableBuilder.append("\n")
-    if (openIndex != closeIndex) {
-      spannableBuilder.setSpan(
-        BulletSpan(), 0, spannableBuilder.length, Spannable.SPAN_INCLUSIVE_EXCLUSIVE
-      )
+  /** Helper marker class. */
+  private class Bullet
+
+  override fun handleOpeningTag(output: Editable) {
+    output.setSpan(Bullet(), output.length, output.length, Spannable.SPAN_INCLUSIVE_EXCLUSIVE)
+  }
+
+  override fun handleClosingTag(output: Editable) {
+    output.append("\n")
+    output.getSpans(0, output.length, Bullet::class.java).lastOrNull()?.let {
+      val start = output.getSpanStart(it)
+      output.removeSpan(it)
+      if (start != output.length) {
+        output.setSpan(BulletSpan(), start, output.length, Spanned.SPAN_INCLUSIVE_EXCLUSIVE)
+      }
     }
-    output.replace(openIndex, closeIndex, spannableBuilder)
   }
 }

--- a/utility/src/main/java/org/oppia/util/parser/CustomHtmlContentHandler.kt
+++ b/utility/src/main/java/org/oppia/util/parser/CustomHtmlContentHandler.kt
@@ -94,6 +94,7 @@ class CustomHtmlContentHandler private constructor(
           currentTrackedCustomTags += TrackedCustomTag(
             localCurrentTrackedTag.tag, localCurrentTrackedTag.attributes, output.length
           )
+          customTagHandlers.getValue(tag).handleOpeningTag(output)
         }
       }
       tag in customTagHandlers -> {
@@ -105,6 +106,7 @@ class CustomHtmlContentHandler private constructor(
           "Expected tracked tag $currentTrackedTag to match custom tag: $tag"
         }
         val (_, attributes, openTagIndex) = currentTrackedCustomTag
+        customTagHandlers.getValue(tag).handleClosingTag(output)
         customTagHandlers.getValue(tag).handleTag(attributes, openTagIndex, output.length, output)
       }
     }
@@ -122,12 +124,32 @@ class CustomHtmlContentHandler private constructor(
     /**
      * Called when a custom tag is encountered. This is always called after the closing tag.
      *
-     * @param attributes The tag's attributes
-     * @param openIndex The index in the output [Editable] at which this tag begins
-     * @param closeIndex The index in the output [Editable] at which this tag ends
-     * @param output The destination [Editable] to which spans can be added
+     * @param attributes the tag's attributes
+     * @param openIndex the index in the output [Editable] at which this tag begins
+     * @param closeIndex the index in the output [Editable] at which this tag ends
+     * @param output the destination [Editable] to which spans can be added
      */
-    fun handleTag(attributes: Attributes, openIndex: Int, closeIndex: Int, output: Editable)
+    fun handleTag(attributes: Attributes, openIndex: Int, closeIndex: Int, output: Editable) {}
+
+    /**
+     * Called when the opening of a custom tag is encountered. This does not support processing
+     * attributes of the tag--[handleTag] should be used, instead.
+     *
+     * This function will always be called before [handleClosingTag].
+     *
+     * @param output the destination [Editable] to which spans can be added
+     */
+    fun handleOpeningTag(output: Editable) {}
+
+    /**
+     * Called when the closing of a custom tag is encountered. This does not support processing
+     * attributes of the tag--[handleTag] should be used, instead.
+     *
+     * This function will always be called before [handleClosingTag].
+     *
+     * @param output the destination [Editable] to which spans can be added
+     */
+    fun handleClosingTag(output: Editable) {}
   }
 
   companion object {

--- a/utility/src/test/java/org/oppia/util/parser/CustomHtmlContentHandlerTest.kt
+++ b/utility/src/test/java/org/oppia/util/parser/CustomHtmlContentHandlerTest.kt
@@ -78,6 +78,26 @@ class CustomHtmlContentHandlerTest {
   }
 
   @Test
+  fun testParseHtml_withOneCustomTag_nonAttributeHandlersAreCalledBeforeAttributeAndInOrder() {
+    val fakeTagHandler = FakeTagHandler()
+
+    CustomHtmlContentHandler.fromHtml(
+      html = "<custom-tag custom-attribute=\"value\">content</custom-tag>",
+      imageGetter = mockImageGetter,
+      customTagHandlers = mapOf("custom-tag" to fakeTagHandler)
+    )
+
+    assertThat(fakeTagHandler.handleOpeningTagCalled).isTrue()
+    assertThat(fakeTagHandler.handleClosingTagCalled).isTrue()
+    assertThat(fakeTagHandler.handleTagCalled).isTrue()
+    // Call order: opening tag -> close tag -> handle tag.
+    assertThat(fakeTagHandler.handleOpeningTagCallIndex)
+      .isLessThan(fakeTagHandler.handleClosingTagCallIndex)
+    assertThat(fakeTagHandler.handleClosingTagCallIndex)
+      .isLessThan(fakeTagHandler.handleTagCallIndex)
+  }
+
+  @Test
   fun testParseHtml_withOneCustomTag_missingHandler_keepsContent() {
     val parsedHtml =
       CustomHtmlContentHandler.fromHtml(
@@ -146,7 +166,13 @@ class CustomHtmlContentHandlerTest {
 
   private class FakeTagHandler : CustomHtmlContentHandler.CustomTagHandler {
     var handleTagCalled = false
+    var handleTagCallIndex = -1
+    var handleOpeningTagCalled = false
+    var handleOpeningTagCallIndex = -1
+    var handleClosingTagCalled = false
+    var handleClosingTagCallIndex = -1
     lateinit var attributes: Attributes
+    private var methodCallCount = 0
 
     override fun handleTag(
       attributes: Attributes,
@@ -155,7 +181,18 @@ class CustomHtmlContentHandlerTest {
       output: Editable
     ) {
       handleTagCalled = true
+      handleTagCallIndex = methodCallCount++
       this.attributes = attributes
+    }
+
+    override fun handleOpeningTag(output: Editable) {
+      handleOpeningTagCalled = true
+      handleOpeningTagCallIndex = methodCallCount++
+    }
+
+    override fun handleClosingTag(output: Editable) {
+      handleClosingTagCalled = true
+      handleClosingTagCallIndex = methodCallCount++
     }
   }
 

